### PR TITLE
Add vod hls

### PR DESCRIPTION
**Pull request by Karolk99**
## Acknowledging the stipulations set forth:
 - [x] I hereby confirm that a Pull Request involving updates to the Software Development Kit (SDK) has been smoothly merged, currently awaits processing, or is otherwise deemed unnecessary in this context.
 - [x] I also affirm that another Pull Request, specifically addressing updates to the documentation body (commonly referred to as 'docs'), has either been successfully incorporated, is in the process of review, or is considered superfluous under the prevailing circumstances.

Before implementing SDK and documentation PRs I want your feedback on these changes

### CODE REVIEW

File: lib/jellyfish/component/hls/recording.ex [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    root_directory()
+    |> File.ls!()
+    |> Enum.filter(&exists?(&1))
+  end
+
+  @spec delete(Room.id()) :: :ok
+  def delete(id) do
+    directory(id) |> File.rm_rf!()
+    :ok
+  end
```
Comment (Rados13) [inline lib/jellyfish/component/hls/recording.ex:29]: Why do we specify this `:ok`?

File: lib/jellyfish_web/controllers/recording_controller.ex [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+          do: put_resp_content_type(conn, @playlist_content_type, nil),
+          else: conn
+
+      Conn.send_resp(conn, 200, file)
+    else
+      {:error, :recording_not_found} ->
+        {:error, :not_found, "Recording not found"}
+
+      {:error, _reason} ->
+        {:error, :not_found, "File not found"}
```
Comment (Rados13) [inline lib/jellyfish_web/controllers/recording_controller.ex:71]: Are there any files that aren't part of the recording?

File: lib/jellyfish_web/controllers/recording_controller.ex [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    |> render("show.json", recordings: recordings)
+  end
+
+  def delete(conn, %{"recording_id" => recording_id}) do
+    if Recording.exists?(recording_id) do
+      Recording.delete(recording_id)
+      send_resp(conn, :no_content, "")
+    else
+      {:error, :not_found, "Recording not found"}
+    end
```
Comment (Rados13) [inline lib/jellyfish_web/controllers/recording_controller.ex:89]: What do you think about that we would always try to delete the recording and if it is successful we send a response `no_content` if fails we return `:not_found`?

File: test/jellyfish/component/hls/request_handler_test.exs [hunk 4]
```diff
@@ -46,15 +46,41 @@ defmodule Jellyfish.Component.HLS.RequestHandlerTest do
...
+
+    # When a room or HLS endpoint finishes, the corresponding path is removed from ETS.
+    # This implies that from this point the HLS stream becomes a recording.
+    EtsHelper.delete_hls_folder_path(room_id)
+
+    assert {:ok, @manifest_content} == RequestHandler.handle_recording_request(room_id, @manifest)
+    assert {:error, :room_not_found} == RequestHandler.handle_file_request(room_id, @manifest)
+
+    assert {:ok, @manifest_content}
   end
```
Comment (Rados13) [inline test/jellyfish/component/hls/request_handler_test.exs:84]: Maybe we could move it into the `describe` block where we would define another setup function, where we would `add_mock_manifest` and `on_exit` remove it?

File: test/jellyfish_web/controllers/component_controller_test.exs [hunk 5]
```diff
@@ -99,6 +104,10 @@ defmodule JellyfishWeb.ComponentControllerTest do
                json_response(conn, :created)
 
       assert_response_schema(response, "ComponentDetailsResponse", @schema)
+      assert_hls_path(room_id, persistent: true)
+
+      # It is persistent stream so we have to remove it manually
+      room_id |> HLS.Recording.directory() |> File.rm_rf!()
```
Comment (Rados13) [inline test/jellyfish_web/controllers/component_controller_test.exs:110]: Could we assert that it is successfully removed?

File: lib/jellyfish/application.ex [hunk 6]
```diff
@@ -41,6 +41,7 @@ defmodule Jellyfish.Application do
         end
 
     :ets.new(:rooms_to_tables, [:public, :set, :named_table])
+    :ets.new(:hls_folder_path, [:public, :set, :named_table])
```
[the remark this answers is not in the document]
Comment (Karolk99) [reply lib/jellyfish/application.ex:44]: `%{Room.id() => path()}`. I can add comment if that would help

File: lib/jellyfish/component/hls/recording.ex [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+  @spec directory(Room.id()) :: String.t()
+  def directory(id) do
+    root_directory() |> Path.join(id)
+  end
+
+  defp root_directory() do
+    base_path = Application.fetch_env!(:jellyfish, :output_base_path)
+    Path.join([base_path, "recordings"])
```
Comment (sgfn) [inline lib/jellyfish/component/hls/recording.ex:38]: This is used only here, but maybe we should make a module attribute anyway? `@recordings_folder` or something?

File: lib/jellyfish/room.ex [hunk 8]
```diff
@@ -420,18 +421,34 @@ defmodule Jellyfish.Room do
   defp hls_component(%{components: components}),
     do:
       Enum.find_value(components, fn {_id, component} ->
-        if component.type == Component.HLS, do: component
+        if component.type == HLS, do: component
```
Comment (sgfn) [inline lib/jellyfish/room.ex:424]: If we're changing it, let's change every occurence in this file

File: lib/jellyfish/room.ex [hunk 9]
```diff
@@ -461,4 +478,10 @@ defmodule Jellyfish.Room do
 
   defp hls_component_already_present?(components),
     do: components |> Map.values() |> Enum.any?(&(&1.type == Component.HLS))
+
+  defp create_folder(directory) do
+    File.rm_rf!(directory)
+    File.mkdir_p!(directory)
+    directory
+  end
```
[the remark this answers is not in the document]
Comment (Karolk99) [reply lib/jellyfish/room.ex]: It was removed anyway so problem solved I guess

**Review by sgfn — COMMENTED** [review]
If everything outlined by @Rados13 has been accounted for, LGTM

Time for the docs 📘  and the SDK 👨‍💻

File: test/jellyfish_web/controllers/component_controller_test.exs [hunk 10]
```diff
@@ -107,7 +107,7 @@ defmodule JellyfishWeb.ComponentControllerTest do
       assert_hls_path(room_id, persistent: true)
 
       # It is persistent stream so we have to remove it manually
-      room_id |> HLS.Recording.directory() |> File.rm_rf!()
+      {:ok, _removed_files} = room_id |> HLS.Recording.directory() |> File.rm_rf()
```
Comment (Rados13) [inline test/jellyfish_web/controllers/component_controller_test.exs:110]:
```suggestion
      assert {:ok, _removed_files} = room_id |> HLS.Recording.directory() |> File.rm_rf()
```

File: lib/jellyfish_web/controllers/recording_controller.ex [hunk 11]
```diff
@@ -64,28 +52,29 @@ defmodule JellyfishWeb.RecordingController do
...
 
   def delete(conn, %{"recording_id" => recording_id}) do
-    if Recording.exists?(recording_id) do
-      Recording.delete(recording_id)
+    with :ok <- Recording.delete(recording_id) do
       send_resp(conn, :no_content, "")
     else
-      {:error, :not_found, "Recording not found"}
+      _error ->
+        {:error, :not_found, "Recording not found"}
```
Comment (Rados13) [inline lib/jellyfish_web/controllers/recording_controller.ex:77]: Wouldn't it be simpler this way?
```elixir
case Recording.delete(recording_id) do 
    :ok -> send_resp(conn, :no_content, "")
    :error -> {:error, :not_found, "Recording not found"}
end
```

**Review by Rados13 — APPROVED** [review]

**Review by sgfn — APPROVED** [review]

**Pull request merged**